# Movie Recommendation Project
## Introduction
In the vast landscape of entertainment, finding the perfect movie to watch can be a daunting task. To address this challenge, I have developed a Movie Recommendation project that leverages three powerful recommendation methods: `Popularity-Based Filtering`, `Content-Based Filtering`, and `Collaborative Filtering`.

## Objective
The primary goal of this project is to provide users with personalized movie recommendations based on their preferences, using a combination of different filtering techniques.

## Methods
### Popularity-Based Filtering:
This method recommends movies based on their overall popularity among all users. The recommendation score is calculated using the **Weighted Rating Formula**:

\[WR = (v / (v + m)) * R + (m + (v + m)) * C \]

where:
- \( v \) is the number of votes for the movie.
- \( m \) is the minimum votes required to be listed.
- \( R \) is the average rating of the movie.
- \( C \) is the mean rating across all movies.

This is a simple yet effective approach that gives higher weight to movies with higher ratings and more votes.

### Content-Based Filtering:
Content-Based Filtering recommends movies by analyzing the content and features of each movie. This method takes into account user preferences and suggests movies with similar characteristics to those the user has enjoyed in the past.

### Collaborative Filtering:
Collaborative Filtering identifies patterns in user behavior and recommends movies based on the preferences of users with similar tastes. For this, the project utilizes the [Surprise](http://surpriselib.com/) library, a Python scikit for building and analyzing recommender systems.

The specific modules used for collaborative filtering are:
- [Dataset](https://surprise.readthedocs.io/en/stable/dataset.html) from Surprise: A module to load a dataset.
- [Reader](https://surprise.readthedocs.io/en/stable/reader.html) from Surprise: A module to define the format of the input file.
- [SVD (Singular Value Decomposition)](https://surprise.readthedocs.io/en/stable/matrix_factorization.html#surprise.prediction_algorithms.matrix_factorization.SVD) from Surprise: A collaborative filtering algorithm based on matrix factorization.
- [model_selection](https://surprise.readthedocs.io/en/stable/model_selection.html) from Surprise: A module providing tools for model selection and evaluation.

## Implementation
The project is implemented in a Jupyter Notebook, utilizing Python and the following data science libraries:

- [Pandas](https://pandas.pydata.org/): A powerful and easy-to-use data manipulation and analysis library.
- [NumPy](https://numpy.org/): The fundamental package for scientific computing with Python.
- [scikit-learn](https://scikit-learn.org/stable/): A machine learning library that provides simple and efficient tools for data analysis and modeling.

For content-based filtering, the project uses the [linear_kernel](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.linear_kernel.html) from scikit-learn. The dataset used for training and testing the recommendation models is sourced from [mention the dataset source].


## <span style="color:#004d80">Popularity-Based Filtering</span>

### Importing Panda Library

In [21]:
import pandas as pd

In [22]:
movies = pd.read_csv('movies.csv')

In [23]:
movies.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124


In [24]:
movies.columns

Index(['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count'],
      dtype='object')

### Calculate a weighted rating

WR = (v / (v + m)) * R + (m + (v + m)) * C

v - number of votes for a movie <br>
m - minimum number of votes required<br>
R - average rating of votes required<br>
C - average rating across all movies


In [25]:
m = movies['vote_count'].quantile(0.9) # 0.9 -> 90%
m

1838.4000000000015

* that means 90% of the movies have vote count less than `1838.40` and 10% of movies have a vote count of higher than `1838.40`

In [26]:
C = movies['vote_average'].mean()
C

6.092171559442016

In [27]:
popularity_movies = movies.copy()

In [32]:
def weighted_rating(df, m=m,C=C):
    
    """
    Calculate the weighted rating for each movie using the given formula.

    Parameters:
    - df (DataFrame): DataFrame containing movie information.
    - m (float): Minimum votes required for a movie to be considered.
    - C (float): Mean rating across all movies.

    Returns:
    - Series: Series containing the calculated 'weighted_rating' for each movie.
    """
    # Extract relevant columns
    R = df['vote_average']
    v = df['vote_count']
    
    # Apply the weighted rating formula
    wr = (v / (v + m)) * R + (m + (v + m)) * C
    
    return wr

In [33]:
popularity_movies['weighted_rating'] = popularity_movies.apply(weighted_rating, axis=1)

In [34]:
popularity_movies.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,...,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count,weighted_rating
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...",...,2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800,94293.550261
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...",...,2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500,49819.367120
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...",...,2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466,49611.797457
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...",...,2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106,77881.333990
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]",...,2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124,35342.738618


In [35]:
# Sorting movies by 'weighted_rating' in descending order and selecting the top 10.
popularity_movies.sort_values('weighted_rating', ascending=False)[['title','weighted_rating']].head(10)

,title,weighted_rating
96,Inception,106186.384534
65,The Dark Knight,95525.050252
0,Avatar,94293.550261
16,The Avengers,94147.509426
788,Deadpool,89389.462627
95,Interstellar,88610.252702
287,Django Unchained,83931.135742
94,Guardians of the Galaxy,81756.277589
426,The Hunger Games,80006.955265
127,Mad Max: Fury Road,79836.622713


### <span style="color:red">Conclusion for Popularity-Based Filtering

* The Popularity-Based Filtering method was employed to recommend movies based on their overall popularity. The weighted rating formula, considering both the average rating and the number of votes, was utilized to prioritize movies that not only received high ratings but also garnered a substantial number of votes.

* The top 10 movies recommended by this approach, as displayed above, showcase the most popular and well-received movies according to the weighted rating criterion. Users looking for widely acclaimed and widely watched movies can refer to this list for their next cinematic experience.
    
    
    
<br><br><br>
    


## <span style="color:#004d80">Content Based Filtering
* focus on `overview` column

In [36]:
movies.head()

,budget,genres,homepage,id,keywords,original_language,original_title,overview,popularity,production_companies,production_countries,release_date,revenue,runtime,spoken_languages,status,tagline,title,vote_average,vote_count
0,237000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.avatarmovie.com/,19995,"[{""id"": 1463, ""name"": ""culture clash""}, {""id"":...",en,Avatar,"In the 22nd century, a paraplegic Marine is di...",150.437577,"[{""name"": ""Ingenious Film Partners"", ""id"": 289...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2009-12-10,2787965087,162.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}, {""iso...",Released,Enter the World of Pandora.,Avatar,7.2,11800
1,300000000,"[{""id"": 12, ""name"": ""Adventure""}, {""id"": 14, ""...",http://disney.go.com/disneypictures/pirates/,285,"[{""id"": 270, ""name"": ""ocean""}, {""id"": 726, ""na...",en,Pirates of the Caribbean: At World's End,"Captain Barbossa, long believed to be dead, ha...",139.082615,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2007-05-19,961000000,169.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"At the end of the world, the adventure begins.",Pirates of the Caribbean: At World's End,6.9,4500
2,245000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://www.sonypictures.com/movies/spectre/,206647,"[{""id"": 470, ""name"": ""spy""}, {""id"": 818, ""name...",en,Spectre,A cryptic message from Bond’s past sends him o...,107.376788,"[{""name"": ""Columbia Pictures"", ""id"": 5}, {""nam...","[{""iso_3166_1"": ""GB"", ""name"": ""United Kingdom""...",2015-10-26,880674609,148.0,"[{""iso_639_1"": ""fr"", ""name"": ""Fran\u00e7ais""},...",Released,A Plan No One Escapes,Spectre,6.3,4466
3,250000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 80, ""nam...",http://www.thedarkknightrises.com/,49026,"[{""id"": 849, ""name"": ""dc comics""}, {""id"": 853,...",en,The Dark Knight Rises,Following the death of District Attorney Harve...,112.312950,"[{""name"": ""Legendary Pictures"", ""id"": 923}, {""...","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-07-16,1084939099,165.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,The Legend Ends,The Dark Knight Rises,7.6,9106
4,260000000,"[{""id"": 28, ""name"": ""Action""}, {""id"": 12, ""nam...",http://movies.disney.com/john-carter,49529,"[{""id"": 818, ""name"": ""based on novel""}, {""id"":...",en,John Carter,"John Carter is a war-weary, former military ca...",43.926995,"[{""name"": ""Walt Disney Pictures"", ""id"": 2}]","[{""iso_3166_1"": ""US"", ""name"": ""United States o...",2012-03-07,284139100,132.0,"[{""iso_639_1"": ""en"", ""name"": ""English""}]",Released,"Lost in our world, found in another.",John Carter,6.1,2124


### Text Feature Extraction using TF-IDF Vectorizer

In [37]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [38]:
tfidf = TfidfVectorizer(stop_words='english')

In [39]:
movies['overview']=movies['overview'].fillna('')

In [40]:
tfidf_matrix = tfidf.fit_transform(movies['overview'])

In [41]:
# Convert TF-IDF Matrix to Pandas DataFrame
pd.DataFrame(tfidf_matrix.toarray(), columns=tfidf.get_feature_names_out())

,00,000,007,07am,10,100,1000,101,108,10th,...,zuckerberg,zula,zuzu,zyklon,æon,éloigne,émigré,été,única,über
0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4798,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4799,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4800,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4801,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [42]:
tfidf_matrix.shape

(4803, 20978)

### Similarity matrix

In [43]:
# Import Linear Kernel for Similarity Calculation
from sklearn.metrics.pairwise import linear_kernel

In [44]:
similarity_matrix = linear_kernel(tfidf_matrix, tfidf_matrix)

In [45]:
similarity_matrix

array([[1.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.        , 0.        , ..., 0.02160533, 0.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.01488159, 0.        ,
        0.        ],
       ...,
       [0.        , 0.02160533, 0.01488159, ..., 1.        , 0.01609091,
        0.00701914],
       [0.        , 0.        , 0.        , ..., 0.01609091, 1.        ,
        0.01171696],
       [0.        , 0.        , 0.        , ..., 0.00701914, 0.01171696,
        1.        ]])

In [46]:
similarity_matrix.shape

(4803, 4803)

### Find the most similar movies to a certain movie

In [47]:
def similar_movies(movie_title, nr_movies):
    """
    Get a list of similar movies based on the cosine similarity between movie vectors.

    Parameters:
    - movie_title (str): The title of the movie for which similar movies are to be found.
    - nr_movies (int): The number of similar movies to retrieve.

    Returns:
    - List[str]: A list of titles of similar movies.
    """
    # Find the index of the movie in the DataFrame
    idx = movies.loc[movies['title']==movie_title].index[0]
    
    # Get the similarity scores for the given movie
    scores = list(enumerate(similarity_matrix[idx]))
    
    # Sort the movies based on similarity scores in descending order
    scores = sorted(scores, key=lambda x: x[1], reverse=True)
    
    # Extract the indices of similar movies (excluding the input movie)
    movies_indices = [tpl[0] for tpl in scores[1:nr_movies+1]]
    
    # Get the titles of similar movies
    similar_titles = list(movies['title'].iloc[movies_indices])
    
    return similar_titles

In [48]:
similar_movies('The Dark Knight',3)

['The Dark Knight Rises',
 'Batman Returns',
 'Batman: The Dark Knight Returns, Part 2']

### <span style="color:red">Conclusion for Content-Based Filtering</span>

* Content-Based Filtering, implemented through the `similar_movies` function, employs cosine similarity scores based on TF-IDF vectors to recommend movies with similar content features to a given movie.

* In this demonstration, movies similar to 'The Dark Knight' have been identified, showcasing the capability of the Content-Based Filtering approach to provide personalized recommendations based on content characteristics.

* This method is particularly valuable for suggesting movies that share thematic elements, genres, or content features, delivering a personalized viewing experience aligned with the user's preferences.

    
<br><br><br>

## <span style="color:#004d80">Collaborative Filtering

In [50]:
ratings = pd.read_csv("ratings.csv")
ratings.head()

,userId,movieId,rating,timestamp
0,1,31,2.5,1260759144
1,1,1029,3.0,1260759179
2,1,1061,3.0,1260759182
3,1,1129,2.0,1260759185
4,1,1172,4.0,1260759205


### Create the dataset

In [51]:
# Importing Dataset and Reader classes from the Surprise library
from surprise import Dataset, Reader

In [52]:
# Creating a Reader instance with the rating scale defined as (1, 5)
reader = Reader(rating_scale=(1,5))

# Loading the dataset from the 'ratings' DataFrame using the specified Reader
dataset = Dataset.load_from_df(ratings[['userId','movieId','rating']],reader)
dataset

### Build the trainset

In [53]:
trainset = dataset.build_full_trainset()

In [54]:
list(trainset.all_ratings())

[(0, 0, 2.5),
 (0, 1, 3.0),
 (0, 2, 3.0),
 (0, 3, 2.0),
 (0, 4, 4.0),
 (0, 5, 2.0),
 (0, 6, 2.0),
 (0, 7, 2.0),
 (0, 8, 3.5),
 (0, 9, 2.0),
 (0, 10, 2.5),
 (0, 11, 1.0),
 (0, 12, 4.0),
 (0, 13, 4.0),
 (0, 14, 3.0),
 (0, 15, 2.0),
 (0, 16, 2.0),
 (0, 17, 2.5),
 (0, 18, 1.0),
 (0, 19, 3.0),
 (1, 20, 4.0),
 (1, 21, 5.0),
 (1, 22, 5.0),
 (1, 23, 4.0),
 (1, 24, 4.0),
 (1, 25, 3.0),
 (1, 26, 3.0),
 (1, 27, 4.0),
 (1, 28, 3.0),
 (1, 29, 5.0),
 (1, 30, 4.0),
 (1, 31, 3.0),
 (1, 32, 3.0),
 (1, 33, 3.0),
 (1, 34, 3.0),
 (1, 35, 3.0),
 (1, 36, 3.0),
 (1, 37, 5.0),
 (1, 38, 1.0),
 (1, 39, 3.0),
 (1, 40, 3.0),
 (1, 41, 3.0),
 (1, 42, 4.0),
 (1, 43, 4.0),
 (1, 44, 5.0),
 (1, 45, 5.0),
 (1, 46, 3.0),
 (1, 47, 4.0),
 (1, 48, 3.0),
 (1, 49, 4.0),
 (1, 50, 3.0),
 (1, 51, 4.0),
 (1, 52, 2.0),
 (1, 53, 1.0),
 (1, 54, 3.0),
 (1, 55, 4.0),
 (1, 56, 4.0),
 (1, 57, 3.0),
 (1, 58, 3.0),
 (1, 59, 3.0),
 (1, 60, 3.0),
 (1, 61, 2.0),
 (1, 62, 3.0),
 (1, 63, 3.0),
 (1, 64, 3.0),
 (1, 65, 3.0),
 (1, 66, 2.0),
 (1, 

### Train the Model

In [55]:
from surprise import SVD

In [56]:
svd = SVD()

In [57]:
svd.fit(trainset)

In [58]:
# Making a prediction for user with ID 15 and item with ID 1956
svd.predict(15,1956)

Prediction(uid=15, iid=1956, r_ui=None, est=3.3127737105813835, details={'was_impossible': False})

* The `est` field in the output represents the estimated rating for the given user-item pair.

### Validation

In [59]:
from surprise import model_selection

In [60]:
# Using cross-validation to evaluate the SVD model's performance on the dataset
model_selection.cross_validate(svd,dataset,measures=['RMSE','MAE'])

{'test_rmse': array([0.89213304, 0.89616525, 0.89376161, 0.9000934 , 0.90307484]),
 'test_mae': array([0.68682168, 0.69086371, 0.68845708, 0.69360254, 0.69470127]),
 'fit_time': (1.518972396850586,
  1.2473247051239014,
  1.220029354095459,
  1.2253117561340332,
  1.2086174488067627),
 'test_time': (0.5035138130187988,
  0.18817400932312012,
  0.18307781219482422,
  0.18808746337890625,
  0.16300082206726074)}

* RMSE: An RMSE of around 0.89 implies that, on average, the predicted ratings differ from the true ratings by approximately 0.89 units. This could be considered acceptable, depending on the scale of your ratings.

* MAE: An MAE of around 0.69 means that, on average, the absolute difference between predicted and true ratings is approximately 0.69. Again, this depends on the scale of your ratings, but a lower MAE is generally desirable.

### <span style="color:red">Conclusion for Collaborative Filtering</span>

* Collaborative Filtering, implemented through the Surprise library and the SVD algorithm, leverages user-item interaction patterns to make personalized movie recommendations.

* The SVD model was trained on the provided dataset, and cross-validation results indicate reasonably good performance with RMSE values around 0.89 and MAE values around 0.69. However, the assessment should be considered in the context of the specific characteristics and requirements of the recommendation system and dataset.

* It's crucial to recognize that the effectiveness of collaborative filtering depends on the availability and quality of user-item interactions. Sparse data, user-item cold-start problems, and the scalability of the model to large datasets should be taken into consideration.

* Collaborative Filtering excels in capturing latent features and uncovering user preferences based on their interactions. It offers a valuable approach for providing personalized recommendations, particularly when content-based information is limited.

    
